core/src/main/java/com/linecorp/armeria/internal/common/stream/HttpDataDeframer.java

```diff
@@ -0,0 +1,417 @@
+                    case 1:
+                        value |= buf.readUnsignedByte();
+                        break;
+                    case 2:
+                        value |= buf.readUnsignedShort();
+                        break;
+                    case 3:
+                        value |= buf.readUnsignedMedium();
+                        break;
+                    default:
+                        throw new Error(); // Should not reach here.
+                }
+                remained -= readSize;
+                if (remained == 0) {
+                    break;
+                }
+            }
+
+            if (remained == 0) {
+                return value;
+            } else {
+                throw new IndexOutOfBoundsException(
+                        "readerIndex: " + (readerIndex() - 4 + remained) +
+                        " (expected: 0 <= readerIndex + 4 <= length(" + length() + "))");
+            }
+        }
+
+        /**
+         * Returns a newly retained slice of this {@link ByteBuf} at the current {@code readerIndex}
+         * and increases the {@code readerIndex} by the specified {@code length}.
+         *
+         * @throws IndexOutOfBoundsException if the specified {@code length} is greater than {@link #length()}
+         */
+        public ByteBuf readBytes(int length) {
+            // fast path
+            final ByteBuf firstBuf = queue.peek();
+            if (firstBuf.readableBytes() >= length) {
+                return firstBuf.readRetainedSlice(length);
+            }
+
+            // slow path
+            ByteBuf value = null;
+            int remained = length;
+            for (ByteBuf buf : queue) {
+                if (!buf.isReadable()) {
+                    continue;
+                }
+
+                final int readSize = Math.min(remained, buf.readableBytes());
+                if (readSize == length) {
+                    return buf.readRetainedSlice(readSize);
+                } else {
+                    if (value == null) {
+                        value = alloc.buffer(length);
+                    }
+                    value.writeBytes(buf, readSize);
+                    remained -= readSize;
+                    if (remained == 0) {
+                        break;
+                    }
+                }
+            }
+
+            if (remained > 0) {
+                ReferenceCountUtil.release(value);
+                throw new IndexOutOfBoundsException(
+                        "readerIndex: " + (readerIndex() - length + remained) +
+                        " (expected: 0 <= readerIndex + " + length + " <= length(" + length() + "))");
+            }
+
+            return value;
+        }
+
+        /**
+         * Returns a byte at the specified absolute {@code index} in this {@link HttpDataReader}.
+         * This method does not modify {@code readerIndex} of this {@link HttpDataReader}.
+         */
+        public byte getByte(int index) {
```

Comment:
Do we need `get*()` methods that do not move the `readerIndex` forward? Could you explain why?

Reply:
This pattern is used to detect the boundary of multipart.
I thought this HttpDeframer could be used in `MimeParser`
https://github.com/ikhoon/armeria/blob/multipart/core/src/main/java/com/linecorp/armeria/common/multipart/MimeParser.java#L338-L350

Reply:
By the way, the MimeParser logic is pretty complex. It would be better to use ByteBuf directly then `HttpDeframeInput`. Let me remove `get*` from this.